optuna/storages/base.py

```diff
@@ -24,162 +24,848 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
+        """
 
         raise NotImplementedError
 
     @abc.abstractmethod
-    def get_all_study_summaries(self):
-        # type: () -> List[study.StudySummary]
+    def get_all_study_summaries(self) -> List[study.StudySummary]:
+
+        """Returns a list of `study.StudySummary` objects.
+
+        Returns:
+            A list of `study.StudySummary` objects.
+
+        """
 
         raise NotImplementedError
 
     # Basic trial manipulation
 
     @abc.abstractmethod
-    def create_new_trial(self, study_id, template_trial=None):
-        # type: (int, Optional[FrozenTrial]) -> int
+    def create_new_trial(self, study_id: int, template_trial: Optional[FrozenTrial]=None) -> int:
+
+        """Create and add a new trial to a specified study.
+
+        Args:
+            study_id:
+                Study ID to add a new trial.
+            template_trial:
+                Fronzen trial with default user-attributes, system-attributes,
+                intermediate-values, and a state.
+
+        Returns:
+            Trial ID of the created trial.
+
+        Raises (current implementation):
+            in-memory:
+                ValueError:
+                    If no study with the matching `study_id` exists.
+            rdb:
+                sqlalchemy.exc.IntegrityError:
+                    If no study with the matching `study_id` exists.
+            redis:
+                AssertionError:
+                    If no study with the matching `study_id` exists.
+
+        Raises (proposal):
+            KeyError:
+                If no study with the matching `study_id` exists.
+        """
 
         raise NotImplementedError
 
     @abc.abstractmethod
-    def set_trial_state(self, trial_id, state):
-        # type: (int, TrialState) -> bool
+    def set_trial_state(self, trial_id: int, state: TrialState) -> bool:
+
+        """Update a state of a specified trial.
+
+        Args:
+            trial_id:
+                Trial ID of a trial to update its state.
+            state:
+                New state of the trial.
+
+        Returns:
+
+        Raise (current implementation):
+            in-memory:
+                IndexError:
+                    If no trial with the matching `trial_id` exists.
+                RuntimeError:
+                    If the trial is already finished.
+            rdb:
+                ValueError:
+                    If no trial with the matching `trial_id` exists.
+                RuntimeError:
+                    If the trial is already finished.
+            redis:
+                AssertionError:
+                    If no trial with the matching `trial_id` exists.
+                RuntimeError:
+                    If the trial is already finished.
+
+        Raises (proposal):
+            KeyError:
+                If no trial with the matching `trial_id` exists.
+            RuntimeError:
+                If the trial is already finished.
```

Comment:
FYI, `study._pop_waiting_trial_id()` calls `set_trial_state()` to get `WAITING` trials. To avoid that multiple workers pop the same trial id, Optuna checked its returned value.

* `True`: This means that the `WAITING` state is properly changed to `RUNNING`.
* `False`: This means that the state has been already `RUNNING` (another worker pop this trial).

This is originally implemented by me. But I'm sorry that it's not clear.
We need to consider another interfaces like:

1. Raise an exception instead of return `False`.
2. Add new storage API for `enqueue_trial()`.
3. Keep the current design (not changed).

Reply:
Thank you for catching the missing doc on the return value, that's my mistake.
I think we need further discussions on the return values of this and other methods.

If getting waiting trials is the only usage of this method, I think it'll better to add `states` argument to `get_all_trials` method.